lib/base/journaldlogger.cpp

```diff
@@ -0,0 +1,81 @@
+/* Icinga 2 | (c) 2021 Icinga GmbH | GPLv2+ */
+
+#include "base/i2-base.hpp"
+#if !defined(_WIN32) && defined(HAVE_SYSTEMD)
+#include "base/journaldlogger.hpp"
+#include "base/journaldlogger-ti.cpp"
+#include "base/configtype.hpp"
+#include "base/statsfunction.hpp"
+#include "base/sysloglogger.hpp"
+#include <sys/uio.h>
+#include <systemd/sd-journal.h>
+
+using namespace icinga;
+
+REGISTER_TYPE(JournaldLogger);
+
+REGISTER_STATSFUNCTION(JournaldLogger, &JournaldLogger::StatsFunc);
+
+void JournaldLogger::StatsFunc(const Dictionary::Ptr& status, const Array::Ptr&)
+{
+	DictionaryData nodes;
+
+	for (const JournaldLogger::Ptr& journaldlogger : ConfigType::GetObjectsByType<JournaldLogger>()) {
+		nodes.emplace_back(journaldlogger->GetName(), 1); //add more stats
+	}
+
+	status->Set("journaldlogger", new Dictionary(std::move(nodes)));
+}
+
+void JournaldLogger::OnConfigLoaded()
+{
+	ObjectImpl<JournaldLogger>::OnConfigLoaded();
+	m_Facility = SyslogHelper::FacilityToNumber(GetFacility());
+}
+
+void JournaldLogger::ValidateFacility(const Lazy<String>& lvalue, const ValidationUtils& utils)
+{
+	ObjectImpl<JournaldLogger>::ValidateFacility(lvalue, utils);
+	if (!SyslogHelper::ValidateFacility(lvalue()))
+		BOOST_THROW_EXCEPTION(ValidationError(this, { "facility" }, "Invalid facility specified."));
+}
+
+/**
+ * Processes a log entry and outputs it to journald.
+ *
+ * @param entry The log entry.
+ */
+void JournaldLogger::ProcessLogEntry(const LogEntry& entry)
+{
+	std::vector<String> sdFields {
+		String("MESSAGE=") + entry.Message.GetData(),
+		String("PRIORITY=") + Value(SyslogHelper::SeverityToNumber(entry.Severity)),
+		String("SYSLOG_FACILITY=") + Value(m_Facility),
+		String("ICINGA2_FACILITY=") + entry.Facility,
+	};
+	const String identifier = GetIdentifier();
+	if (!identifier.IsEmpty()) {
+		sdFields.push_back(String("SYSLOG_IDENTIFIER=" + identifier));
+	}
+	SystemdJournalSend(sdFields);
+}
+
+void JournaldLogger::Flush()
+{
+	/* Nothing to do here. */
+}
+
+void JournaldLogger::SystemdJournalSend(const std::vector<String>& journalFields)
+{
+	struct iovec iovec[journalFields.size()];
+	int iovecCount = 0;
+
+	for (const String& journalField: journalFields) {
+		iovec[iovecCount++] = (struct iovec) {
+			.iov_base = (char*) journalField.CStr(),
```

Comment:
This actually casts away the const qualifier. Is that safe to do here? And if so, it should probably state what it is, a `const_cast<char*>`.

Reply:
[sd_journal_sendv implementation](https://github.com/systemd/systemd/blob/5efbd0bf897a990ebe43d7dc69141d87c404ac9a/src/libsystemd/sd-journal/journal-send.c#L206) declares it's *iov argument const, and and actually doesn't wirte to the passed buffer. At systemd they're [also casting away const](https://github.com/systemd/systemd/blob/ed056c560b47f84a0aa0289151f4ec91f786d24a/src/basic/io-util.h#L79). Was in doubt on my own, but felt it's enough too cast away the const. Changed into your suggestion.